src/core/src/op/loop.cpp

```diff
@@ -191,6 +191,66 @@ void op::v5::Loop::validate_and_infer_types() {
+                    ov::as_type_ptr<v0::TensorIterator::BodyOutputDescription>(output_description)) {
+                const ov::PartialShape& body_value_shape = body_value.get_partial_shape();
+                if (body_value_shape.rank().is_static()) {
+                    // handle the case: when sub-model's output shape does not compatible input shape in a back-edge,
+                    // such as
+                    //          Parameter(out:-1, 1)->|
+                    //                                |->Concat(out:-1, 2)->Result(out:-1, 2)
+                    //          Parameter(out:-1, 1)->|   (axis==1)
+                    // when iteration number is unknown or sub-model output shape may be vary, the Result shape should
+                    // infer as (-1, -1), then set changed one to input and propagate to others.
+                    if (back_edges.count(output_description->m_body_value_index)) {
+                        auto input_param =
+                            m_bodies[0]->get_parameters().at(back_edges[output_description->m_body_value_index]);
+                        const auto& input_param_ps = input_param->get_partial_shape();
+                        if (input_param_ps.rank().is_static()) {
+                            const auto body_rank_len = body_value_shape.rank().get_length();
+                            const auto input_rank_len = input_param_ps.rank().get_length();
+                            // if input and output shape are {n} or {} and not compatible, make output shape dynamic
+                            if (body_rank_len <= 1 && input_rank_len <= 1 &&
+                                !input_param_ps.compatible(body_value_shape)) {
+                                // reset sub model input shape
+                                input_param->set_partial_shape({-1});
+                                need_reinvalidate = true;
+                            } else if (body_rank_len == input_rank_len &&
+                                       !input_param_ps.compatible(body_value_shape)) {
+                                ov::PartialShape new_ps(body_value_shape);
+                                for (auto i = 0; i < body_value_shape.size(); i++) {
+                                    if (!body_value_shape[i].compatible(input_param_ps[i])) {
+                                        new_ps[i] = Dimension::dynamic();
+                                    }
+                                }
+                                // reset sub model input shape
+                                input_param->set_partial_shape(new_ps);
+                                need_reinvalidate = true;
+                            }
+                        }
+                    }
+                }
+            }
+        }
+        // only input shape changed we will re-compute output shape
+        if (need_reinvalidate) {
+            m_bodies[0]->validate_nodes_and_infer_types();
```

Comment:
Why is this enough for the shape convergence, and finally not underestimated shape bound? As input shapes were modified for the body, then the body can infer different outputs and we should reiterate shape adjustment implemented in the previous block of code again. I don't see this re-iteration.

Reply:
@itikhono, please comment on that as well.

Reply:
@luo-cheng2021 what do you mean by this sentence "The loop is unrolled for clearer logic" ?

Reply:
"As input shapes were modified for the body, then the body can infer different outputs and we should reiterate shape adjustment implemented in the previous block of code again. I don't see this re-iteration."
I guess it means that we need to re-validate the shapes for the body as many times as number of iterations in Loop op, because we assume that the outputs shapes can be changed in each iteration, but now we only do it once. Looks like we need to move lines 204-250 inside some `for ( i = 0; i < num_of_iterations; ...)` loop

Reply:
The number of iterations can be big or unknown. If it is too big we spend a lot of time in this shape inference. Instead we have to save time (because it eventually impacts loading time in the core) with achieving a good compromise between time of the analysis and the accuracy of the resulting estimation. This doesn't look trivial, so we have to apply heuristics. One of the heuristic can be just reiterating the adjustment part. The number of iterations is limited by the sum of all input/output tensor ranks (kind of) because each time when shape is changed, we set at least one dimension to be dynamic and will not revert it to a static one on the next iterations. Does it sound reasonable?

Reply:
Totally agree. Currently worked case has the following heuristic:
1. The Result's rank of the backedge should be same as the Parameter's
2. Each iteration will only change the same dimension's size
3. The changed size of dimension must be seen in `validate_and_infer_types()`
Then 2 pass should be enough, the first pass(Line199~243) will find all changed parameter's shape set new shape(because rule 3 we can find all changes), the second pass (Line245~285) will infer the shape.

Reply:
> @luo-cheng2021 what do you mean by this sentence "The loop is unrolled for clearer logic" ?

My mean is unroll the 2 pass, the original implementation is this: https://github.com/openvinotoolkit/openvino/pull/10684/commits/a4fbe2e448a7f8a7264914722826ed03b5df8cdc

Reply:
Can we generalize this code right now to cover all possible variants without these heuristics?
 I guess it's enough to add this `for ( i = 0; i < num_of_iterations; ...)` loop and check that the body output shapes in the current iteration actually differ from the body output shapes in the previous iteration. If the shapes are the same, we can just break this `for` loop.
Also, as @slyalin mentioned we can speed up the process of re-validation of the shapes because we don't need to validate the whole body ( `m_bodies[0]->validate_nodes_and_infer_types();`) , but only certain part of the body. For example, let's say the shapes of Parameter2 and Parameter3 have changed in iteration N, but the shape of Parameter1 remains the same, so we need to re-validate shapes for operations which actually depends on the Parameter2 and Parameter3.